test/unit/utils.test.ts

```diff
@@ -1165,4 +1170,119 @@ describe('driver utils', function () {
       });
     });
   });
+
+  describe('aborted()', () => {
+    context('when given an aborted signal', () => {
+      it('rejects the returned promise', async () => {
+        const controller = new AbortController();
+        controller.abort(new MongoError('my error'));
+        const error = await aborted(controller.signal).abort.catch(error => error);
+        expect(error).to.be.instanceOf(MongoError);
+        expect(error.message).to.equal('my error');
+      });
+    });
+
+    context('when given a signal that aborts after some time', () => {
+      it('rejects the returned promise after the timeout', async () => {
+        const controller = new AbortController();
+        setTimeout(() => controller.abort(new MongoError('my error')), 10);
+        const start = +new Date();
+        const error = await aborted(controller.signal).abort.catch(error => error);
+        const end = +new Date();
+
+        expect(error).to.be.instanceOf(MongoError);
+        expect(error.message).to.equal('my error');
+        expect(end - start).to.be.within(8, 13);
+      });
+    });
+
+    context('when given a nullish value', () => {
+      it('returns a promise that pends forever', async () => {
+        const error = await Promise.race([
+          aborted().abort.catch(error => error),
+          aborted(AbortSignal.timeout(100)).abort
+        ]).catch(error => error);
+        expect(error.name).to.equal('TimeoutError');
+      });
+    });
+  });
+
+  describe('abortable()', () => {
```

Comment:
we're missing some cases here; based on the functionality we want, the following needs to be covered:
1) already rejected promise with already aborted signal
2) already resolved promise with already aborted signal
3) already rejected promise with not yet aborted signal
4) already resolved promise with not yet aborted signal
5) unresolved promise with an already aborted signal
6) eventually rejecting promise with not yet aborted signal
7) eventually resolving promise with not yet aborted signal
8) unresolved promise with eventually aborted signal

and for eventually rejecting/resolving I would add a timeout to give it a full run of the event loop

Reply:
Added the test cases, thank you for helping lay them out.

2 is a little questionable in the current state - "already resolved promise with already aborted signal". The current behavior I have is that the promise resolution wins, but should that be the case? I think if the signal has already been interrupted it unfortunately doesn't matter that the promise has resolved. Thoughts?

Reply:
The new test cases look good to me now, but would it still be worthwhile to have some additional assertions on cleanup like you did in your original ones?